src/lis3dhtr_spi.cpp

```diff
@@ -0,0 +1,207 @@
+  return hal::success();
+}
+
+hal::status lis3dhtr_spi::configure_data_rates(data_rate_config p_data_rate)
+{
+  constexpr auto configure_reg_bit_mask = hal::bit_mask::from<7, 4>();
+
+  auto read_from_ctrl_reg1 = static_cast<hal::byte>(ctrl_reg1 | 0x80);
+  auto write_to_ctrl_reg1 = static_cast<hal::byte>(ctrl_reg1 | 0x40);
+  auto ctrl_reg1_array = std::array{ read_from_ctrl_reg1 };
+
+  HAL_CHECK(m_cs->level(false));
+  auto ctrl_reg1_data =
+    HAL_CHECK(hal::write_then_read<1>(*m_spi, ctrl_reg1_array));
+  HAL_CHECK(m_cs->level(true));
+
+  hal::bit_modify(ctrl_reg1_data[0])
+    .insert<configure_reg_bit_mask>(static_cast<hal::byte>(p_data_rate));
+
+  HAL_CHECK(m_cs->level(false));
+  HAL_CHECK(
+    hal::write(*m_spi, std::array{ write_to_ctrl_reg1, ctrl_reg1_data[0] }));
+  HAL_CHECK(m_cs->level(true));
+
+  return hal::success();
+}
+
+hal::status lis3dhtr_spi::configure_full_scale(max_acceleration p_gravity_code)
+{
+  m_gscale = static_cast<hal::byte>(p_gravity_code);
+
+  constexpr auto configure_reg_bit_mask = hal::bit_mask::from<5, 4>();
+
+  auto read_from_ctrl_reg4 = static_cast<hal::byte>(ctrl_reg4 | 0x80);
+  auto write_to_ctrl_reg4 = static_cast<hal::byte>(ctrl_reg4 | 0x40);
+
+  auto ctrl_reg4_array = std::array{ read_from_ctrl_reg4 };
+
+  HAL_CHECK(m_cs->level(false));
+  auto ctrl_reg4_data =
+    HAL_CHECK(hal::write_then_read<1>(*m_spi, ctrl_reg4_array));
+  HAL_CHECK(m_cs->level(true));
+
+  hal::bit_modify(ctrl_reg4_data[0])
+    .insert<configure_reg_bit_mask>(static_cast<hal::byte>(p_gravity_code));
+
+  HAL_CHECK(m_cs->level(false));
+  HAL_CHECK(
+    hal::write(*m_spi, std::array{ write_to_ctrl_reg4, ctrl_reg4_data[0] }));
+  HAL_CHECK(m_cs->level(true));
+
+  return hal::success();
+}
+
+// private
+
+lis3dhtr_spi::lis3dhtr_spi(hal::spi& p_spi,
+                           hal::output_pin& p_cs,
+                           max_acceleration p_gscale)
+  : m_spi(&p_spi)
+  , m_cs(&p_cs)
+  , m_gscale(static_cast<hal::byte>(p_gscale))
+{
+}
+
+hal::result<accelerometer::read_t> lis3dhtr_spi::driver_read()
+{
+  accelerometer::read_t acceleration;
+  constexpr auto number_of_axis = 3;
+  constexpr auto bytes_per_axis = 2;
+
+  auto read_from_xyz_register = static_cast<hal::byte>(out_x_l | 0xC0);
+
+  HAL_CHECK(m_cs->level(false));
+  auto xyz_acceleration =
+    HAL_CHECK(hal::write_then_read<number_of_axis * bytes_per_axis>(
+      *m_spi, std::array{ read_from_xyz_register }));
+  HAL_CHECK(m_cs->level(true));
+
+  /* parsing data from accelerometer
+   all data is left justified which means the data will always have the lowest
+   nibble be 0 so we shift it all to the right by 4 and or the low and high
+   bytes together 0000'0000'0000'0000
+  */
+  constexpr auto read_h_bit_mask = hal::bit_mask::from<15, 8>();
+
+  auto x = static_cast<uint16_t>(xyz_acceleration[0]);
+  hal::bit_modify(x).insert<read_h_bit_mask>(
+    static_cast<uint16_t>(xyz_acceleration[1]));
```

Comment:
```suggestion
    const auto x = hal::bit_value(0).insert<low_mask>(xyz_acceleration[0]).insert<high_mask>(
    static_cast<uint16_t>(xyz_acceleration[1]));
```

I'd recommend sticking with the bit manipulation library fully if you can. You can make arbitrary values using hal::bit_value().

Make sure to make these const variables `const`.